### PR TITLE
Fix wide character support

### PR DESCRIPTION
**Pull request by nosami**
Fixes https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1501927

part of https://github.com/xamarin/vsmac/pull/6676

### CODE REVIEW

File: XtermSharp/InputHandlers/InputHandler.cs [hunk 1]
```diff
@@ -1200,56 +1200,75 @@ unsafe void Print (byte* data, int start, int end)
...
+						ch = rune.ToString()[0];
+					}
 
-				// calculate print space
-				// expensive call, therefore we save width in line buffer
+				}
+				// Invalid UTF-8 sequence, client sent us some junk, happens if we run with the wrong locale set
+				// for example if LANG=en
+				else if (readingBuffer.BytesLeft() >= (n - 1)) {
+					var bytes = new byte[n];
```
Comment (Therzok) [inline XtermSharp/InputHandlers/InputHandler.cs:1233]: Wrote on Teams about this as well, please use `ArrayPool<byte>.Shared` (bump to net6.0 or System.Buffers) nuget.

Pattern would be something like:
```csharp
var bytes = ArrayPool<byte>.Rent(n);
try {
   ...
} finally {
    ArrayPool<byte>.Shared.Return(bytes);
}

```

File: XtermSharp/InputHandlers/InputHandler.cs [hunk 2]
```diff
@@ -1200,56 +1200,75 @@ unsafe void Print (byte* data, int start, int end)
...
+							chWidth = 1;
+							chSet = true;
+						}
 					}
-				}
 
-				// MIGUEL-TODO: I suspect this needs to be a stirng in C# to cope with Grapheme clusters
-				var ch = code;
+					if (!chSet) {
+						(var rune, var size) = Rune.DecodeRune(new byte[] { bufferValue });
```
Comment (Therzok) [inline XtermSharp/InputHandlers/InputHandler.cs:1224]: Another candidate for reuse, we can allocate the byte array above the loop here.

File: XtermSharp/InputHandlers/InputHandler.cs [hunk 3]
```diff
@@ -1200,56 +1200,75 @@ unsafe void Print (byte* data, int start, int end)
...
+						chWidth = size;
+						ch = rune.ToString()[0];
```
Comment (Therzok) [inline XtermSharp/InputHandlers/InputHandler.cs:1226]: Completely undiscoverable way of doing this without allocations.

I realize it's out of scope to use [System.Text.Rune](https://docs.microsoft.com/en-us/dotnet/api/system.text.rune?view=net-6.0)

```csharp
var decodeBuf = new byte[1]; // before loop
var encodeBuf = new byte[4];

// loop, etc

if (!chSet) {
    decodeBuf[0] = bufferValue;
    (var rune, var size) = Rune.DecodeRune(decodeBuf);
    chWidth = size;

    // ToString seems to call this.
    Rune.EncodeRune(rune, encodeBuf);
    ch = BitConverter.ToChar(encodeBuf, 0);
}
```

Comment (Therzok) [reply XtermSharp/InputHandlers/InputHandler.cs:1226]: does this only run when we don't have a glyph size calculated for the current text draw? If so, I'm not worried that much, but buffer reuse is always a big win.

**Comment by sandyarmstrong** [comment]
Merging this because we're already using it in VSMac. @nosami please let me know if this needs additional work.

**Comment by sandyarmstrong** [comment]
Not deleting branch to avoid breaking VSMac build.

**Comment by nosami** [comment]
> Merging this because we're already using it in VSMac. @nosami please let me know if this needs additional work.

Oh wow, I forgot all about this. In fact, I thought I'd already finished working on this.

I'm happy for this PR to be merged as-is though. It's not any worse than master and we can revisit at a later date to apply Marius's improvement suggestions.

**Comment by sandyarmstrong** [comment]
Cool, I'll bump VSMac to use `master` again instead of this branch.

**Comment by nosami** [comment]
Thanks Sandy!

**Comment by Therzok** [comment]
> > Merging this because we're already using it in VSMac. @nosami please let me know if this needs additional work.
> 
> Oh wow, I forgot all about this. In fact, I thought I'd already finished working on this.
> 
> I'm happy for this PR to be merged as-is though. It's not any worse than master and we can revisit at a later date to apply Marius's improvement suggestions.

Filed a work item: https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1539938

**Pull request merged**